### PR TITLE
Migrate legacy watchlists to private with rollback inventory

### LINKED ISSUE #8369: [AF-13A/P4] Migrate existing public watchlists to private with rollback runbook
> **Scope reset (2026-09-07):** The authoritative intervention directives in this issue discussion supersede conflicting scope, dependency, sequencing, or pilot language below. Historical text remains for traceability. Every rendered UI/product-copy change remains blocked pending explicit linked human approval; agents may not merge.

Parent: #8348
Epic: #8323
Depends on deployment of #8364, #8367, and #8368. Blocks public cache/index purge.

## Scope

Inventory public state, transactionally make every existing watchlist private without deleting or rewriting user configuration, and retain a tested rollback artifact through the compatibility window. Preserve copy provenance and the dormant future authorization seams for grant/share/template copying and authenticated owner-scoped API/MCP access; only public visibility/discovery is being removed.

Evidence: `apps/web/src/db/schema.ts` still has `isPublic`, partial index `idx_wl_public`, and `sourceWatchlistId`; migration 0066 historically defaulted public, while 0083 changed only the default and deliberately preserved visibility counts. Backup coverage includes `watchlist` and `watchlist_company`.

## Dependencies and rollout gate

- Do not run until owner-only mutations/routes and anonymous REST/MCP cutover are deployed and verified.
- #8364 must prove copy permission no longer depends on `is_public`; #8367 must prove public API/MCP consumers are disabled.
- Independent database and privacy review required before production execution.
- Use a staged runbook with explicit go/no-go, rollback owner, and post-run observation window.
- No rendered UI in this phase; `gate:human-ui` is not applicable.

## Acceptance criteria

- [ ] Preflight inventory records public row IDs, owner/path variants, localized URLs, filters, alerts, companies, AI state if present, `source_watchlist_id`, and aggregate checksums/counts.
- [ ] Backup/restore coverage for `watchlist` and `watchlist_company` is verified before mutation.
- [ ] One reviewed transaction sets existing `is_public=true` rows false without deleting rows.
- [ ] Postconditions prove row, owner, slug, filter, alert, company-membership, AI-state, and provenance counts/checksums are unchanged except visibility.
- [ ] Existing independent copies remain owned by their current users and retain ancestry; no copy or source row is deleted.
- [ ] The migration does not remove future cross-user copy capability: owned works now, future grant/share/template policies remain possible, and unauthorized cross-user mode stays dormant until a real authorization source exists.
- [ ] The migration does not remove future API/MCP capability: anonymous readers stay disabled, while later authenticated callers can be restored only to owner-scoped list/read.
- [ ] Owners can still read every migrated watchlist; anonymous/cross-owner probes are uniformly not-found.
- [ ] The exact pre-migration public-row inventory is retained as rollback input until the rollback window closes.
- [ ] Migration is idempotent or safely detects an already-completed state.
- [ ] Runbook defines abort thresholds, rollback commands, responsible reviewer, telemetry checks, and handoff to the purge phase.
- [ ] `is_public` and `idx_wl_public` remain intact during the rollback window.

### PR DESCRIPTION
**Pull request by viktor-shcherb**
## Summary

- add a guarded, transactional 0089 data migration that snapshots every watchlist and membership, stores the exact legacy-public rollback inventory, and changes only `is_public=true` rows to private
- preserve row/filter/alert/company/source-provenance/owner content with before/after digests and exact set comparisons; retain `is_public` and `idx_wl_public` for rollback compatibility
- capture canonical username and display-username provenance (including mirrored values), slug, all `en`/`de`/`fr`/`it` page/current-OG paths, and the historical localized `opengraph-image-:hash` route plus purge pattern for #8370
- add a direct-unpooled inventory/apply/verify/rollback tool and a staged operator runbook with backup evidence, abort thresholds, rollback owner, observation checks, and purge handoff

## Stack and rollout blockers

This PR was rebased onto `main` after #8376 merged. It is Ready for Review for checkpoint visibility only; do not merge, deploy, or execute it until:

1. #8376 is deployed and verified;
2. #8368 has passed its human UI approval and its owner-only route cutover is deployed and verified; and
3. #8367's anonymous REST/MCP cutover is deployed and its verification run is recorded.

No workflow invokes 0089 in this PR, and no production command was run. The apply path requires session-local evidence for the #8376 and #8368 deployment SHAs, #8368 human approval, a fresh backup/restore run, and the exact reviewed inventory digest. It also fails closed without an immutable #8367 deployment SHA and positive verification-run ID; SQL validates both attested values before locking or mutation.

There is no rendered UI or cost/provider change in this incremental PR, so it has no independent human-UI or cost gate; the #8368 human gate remains a hard dependency.

## Scope exclusions

No route/UI change, copy-authorization expansion, cache/SEO/OG/IndexNow/Typesense purge (#8370), notification behavior, production execution, schema/column/index removal (#8371), or cost change is included.

## Verification

- `pnpm --filter @jobseek/web typecheck` — passed (`next typegen && tsc`)
- `pnpm --filter @jobseek/web exec vitest run src/db/__tests__/watchlist-privacy-interview-migration.test.ts src/db/__tests__/watchlist-visibility-contract.test.ts src/db/__tests__/notification-policy-migration.test.ts` — 3 files / 23 tests passed
- `pnpm --filter @jobseek/web db:check` — passed
- focused ESLint on the migration tool, contract, and tests — passed
- `git diff --check` — passed
- commit and push hooks — passed

No production database or migration command was run.

Closes #8369
Refs #8348
Refs #8376
Refs #8368
Refs #8367

## Remote checkpoint — 2026-09-01

**Ready for Review is a review-visibility state only; it is not merge, deployment, or migration-execution authorization.** This PR remains dependency-blocked.

### Remote and review evidence

- Remote head: `codex/watchlist-visibility-migration` at [`35d4ab0e3cb6d0a41e0cb34981d78f054b6dd2ab`](https://github.com/colophon-group/jobseek/commit/35d4ab0e3cb6d0a41e0cb34981d78f054b6dd2ab).
- The clean local worktree and local branch matched that remote SHA at checkpoint time; no local-only commit or uncommitted PR change was found.
- Exact-head technical approval: [final migration review](https://github.com/colophon-group/jobseek/pull/8381#pullrequestreview-5070770553).
- Current base audit: `main` was `3c20e13c3b97481b9bf5a1d3b894c773c38ad210`; GitHub reported the branch **mergeable with no conflict**, 4 commits ahead and 6 behind ([compare](https://github.com/colophon-group/jobseek/compare/main...codex/watchlist-visibility-migration)). It is behind but mergeable, so this checkpoint did not rewrite/rebase the reviewed head.
- The technical review records passing `git diff --check`, focused Vitest (3 files / 23 tests), targeted ESLint, `pnpm db:check`, and full web typecheck. No production migration command was run.

### Exact dependency and execution blockers

1. #8376 is merged at `bf297095a4610a975984201e3ac0f6d7bd9df90a` and has a [successful production deployment](https://github.com/colophon-group/jobseek/actions/runs/33433820772/job/99625330499), but the migration still requires its exact deployed SHA and verification evidence in the session-local attestation.
2. #8367/#8374 is merged at `2dc8ae8539fde551cdf487b8e73bb55c9d2fc6fc` and has successful production deployment records ([web](https://github.com/colophon-group/jobseek/actions/runs/33425365048/job/99597507467), [crawler](https://github.com/colophon-group/jobseek/actions/runs/33425365289/job/99597507678)); the apply path still fails closed until an immutable deployment SHA and a positive anonymous REST/MCP verification-run ID are supplied.
3. #8385/#8368 is still open. It must receive explicit human taste approval, then be merged, deployed, and verified as the owner-only route cutover before this PR may merge, deploy, or execute. Ready for Review on either PR does not satisfy that dependency.
4. Before execution, record a fresh backup/restore run, the exact reviewed inventory digest, all required session-local attestations, go/no-go approval, and rollback ownership. No production database or migration command has been run.
5. Do not merge this dependency-blocked migration merely because it is technically approved and conflict-free.

### CODE REVIEW

File: apps/web/scripts/watchlist-visibility-migration.ts [hunk 1]
```diff
@@ -0,0 +1,907 @@
...
+      "Public inventory lost an owner or row",
+    );
+    for (const row of publicWatchlists) {
+      const ownerSlugs = new Set(
+        [row.ownerUsername, row.ownerDisplayUsername].filter(
+          (value): value is string => value !== null,
+        ),
+      );
+      invariant(
+        row.pathVariants.length === ownerSlugs.size * 4,
```
Comment (viktor-shcherb) [inline apps/web/scripts/watchlist-visibility-migration.ts]: `readPublicRows` emits one row per slug *kind* (8 variants when both columns are non-null), but this assertion counts distinct slug *values*. Normal account creation and rename mirror `username` and `display_username` to the same value (`src/lib/auth.ts:153`; `src/lib/actions/preferences.ts:422-424`), so an ordinary public owner produces 8 rows while this expects 4 and `inventory` aborts before an artifact can be created. Validate the `(kind, locale)` matrix, or deduplicate the SQL and update the manifest contract consistently. Please add a behavioral test with equal non-null values; the current source-string test cannot catch this.

File: apps/web/scripts/watchlist-visibility-migration.ts [hunk 2]
```diff
@@ -0,0 +1,907 @@
...
+    })}\n`,
+  );
+}
+
+function requiredApplyEvidence(artifact: InventoryArtifact) {
+  const confirmation = process.env.WATCHLIST_PRIVACY_CONFIRMATION;
+  const rawBackupRunId = process.env.WATCHLIST_PRIVACY_BACKUP_RESTORE_RUN_ID;
+  const privateMutationsDeploySha =
+    process.env.WATCHLIST_PRIVATE_MUTATIONS_DEPLOY_SHA;
+  const routeCutoverDeploySha = process.env.WATCHLIST_ROUTE_CUTOVER_DEPLOY_SHA;
```
Comment (viktor-shcherb) [inline apps/web/scripts/watchlist-visibility-migration.ts]: The hard gate says #8367 anonymous REST/MCP cutover must be deployed, but `apply` only requires deploy evidence for #8376 and #8368. Neither `requiredApplyEvidence` nor the temporary/SQL attestation has a #8367 SHA or verification field, so the production write path can run while that dependency is only documented. Add immutable #8367 deploy/verification evidence to the required environment, attestation table/validation, and runbook command.

File: apps/web/scripts/watchlist-visibility-migration.ts [hunk 3]
```diff
@@ -0,0 +1,907 @@
...
+dotenv.config({ path: ".env.local", quiet: true });
+
+type Command = "inventory" | "apply" | "verify" | "rollback";
+
+type PathVariant = {
+  locale: "en" | "de" | "fr" | "it";
+  ownerSlugKind: "username" | "display_username";
+  ownerSlug: string;
+  pagePath: string;
+  ogPath: string;
```
Comment (viktor-shcherb) [inline apps/web/scripts/watchlist-visibility-migration.ts]: `pathVariants` records only the current `/og/watchlist/...` route. The base still accepts the historical Next OG form `/:lang/:userSlug/:watchlistSlug/opengraph-image-:hash` and permanently redirects it (`next.config.ts:152-154`). #8370 requires a removal manifest for every historical watchlist OG URL, so the handoff is incomplete without that legacy form (at least an explicit purge pattern if exact historical hashes cannot be enumerated). Add it to the inventory/runbook and cover both OG shapes.

**Review by viktor-shcherb — COMMENTED** [review]
Blocking technical draft review: three actionable findings remain. GitHub does not permit REQUEST_CHANGES on a self-authored PR, so this is a COMMENT review and is not approval.

The incremental scope is otherwise disciplined: data-only files, a single transactional apply path with migration/table locks, exact row/content digests for watchlists/memberships/filters/alerts/provenance/owners, rollback limited to inventoried IDs, retained `is_public`/`idx_wl_public`, strict migration-ledger drift checks, direct/unpooled write guards, and no workflow or production execution path. The runbook documents the dependency order, but two of those requirements are not yet fully represented in executable evidence/inventory.

Checks run against `2104106`: `git diff --check` passed; focused Vitest passed (1 file, 9 tests); targeted ESLint passed; `pnpm db:check` passed. No production database command was run.

File: apps/web/scripts/watchlist-visibility-migration.ts (hunk 1 above)
Comment (viktor-shcherb) [reply apps/web/scripts/watchlist-visibility-migration.ts]: Fixed in 5c49466b2. Inventory now validates the complete labelled `(ownerSlugKind, locale)` matrix and separately proves each distinct page/current-OG/legacy-OG URL set. Mirrored `username === display_username` retains 8 labelled provenance rows and 4 distinct URLs per form; different values retain 8 distinct URLs. Behavioral tests cover equal values, different values, and a missing labelled variant. Focused suite: 2 files / 15 tests passed.

File: apps/web/scripts/watchlist-visibility-migration.ts (hunk 2 above)
Comment (viktor-shcherb) [reply apps/web/scripts/watchlist-visibility-migration.ts]: Fixed in 5c49466b2. Apply now requires `WATCHLIST_PUBLIC_API_CUTOVER_DEPLOY_SHA` (40-hex deployed #8367 SHA) and `WATCHLIST_PUBLIC_API_CUTOVER_VERIFICATION_RUN_ID` (positive immutable run ID). Both fields are inserted into the session-local attestation and independently validated by 0088 SQL before locks/mutation. Behavioral tests prove valid evidence is bound and each missing #8367 field fails closed; the runbook and draft gate were updated.

File: apps/web/scripts/watchlist-visibility-migration.ts (hunk 3 above)
Comment (viktor-shcherb) [reply apps/web/scripts/watchlist-visibility-migration.ts]: Fixed in 5c49466b2. Every labelled locale/owner variant now carries the current `/og/watchlist/...` path, the historical `/:lang/:userSlug/:watchlistSlug/opengraph-image-:hash` pattern, and an actionable `opengraph-image-*` purge pattern because historical hashes are not reconstructable. Tests cover both OG forms for mirrored and distinct usernames, and the runbook makes this an explicit #8370 handoff requirement.

File: apps/web/scripts/watchlist-visibility-migration.ts [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+
+function writePrivateJson(path: string, value: unknown, exclusive: boolean): void {
+  writeFileSync(resolve(path), `${JSON.stringify(value, null, 2)}\n`, {
+    encoding: "utf8",
+    mode: 0o600,
+    flag: exclusive ? "wx" : "w",
+  });
+}
+
+async function readAggregateEvidence(sql: Sql): Promise<AggregateEvidence> {
```
Comment (viktor-shcherb) [inline apps/web/scripts/watchlist-visibility-migration.ts]: `pnpm typecheck` fails because transaction callbacks pass `TransactionSql` into this and the other query helpers declared as the full `Sql` client. The same TS2345 occurs at runner lines 403, 404, 428, 429, 528, 668, 676-678, 765, 774, and 776; base `52d4fa3` typechecks cleanly. Please type these helpers against the common queryable/transaction interface so the actual transactional call path compiles.

File: apps/web/scripts/watchlist-visibility-contract.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    ),
+    privateMutationsDeploySha: deployedSha(
+      environment,
+      "WATCHLIST_PRIVATE_MUTATIONS_DEPLOY_SHA",
+    ),
+    routeCutoverDeploySha: deployedSha(
+      environment,
+      "WATCHLIST_ROUTE_CUTOVER_DEPLOY_SHA",
+    ),
+    routeCutoverApprovedBy,
```
Comment (viktor-shcherb) [inline apps/web/scripts/watchlist-visibility-contract.ts:186]: `routeCutoverApprovedBy` remains inferred as `string | undefined`: the optional-chain truthiness assertion above does not narrow the original variable, so interpolation into the attestation insert produces TS2769 at `watchlist-visibility-migration.ts:544`. Return a definitely narrowed non-empty string here; this head must pass the package typecheck before technical approval.

**Review by viktor-shcherb — COMMENTED** [review]
Follow-up technical review: the three original blockers are resolved, but this head is not ready for technical approval because the web package typecheck now fails. GitHub does not permit REQUEST_CHANGES on this self-authored PR, so this is a blocking COMMENT review rather than approval.

Verified at `5c49466`: equal mirrored slugs retain 8 labelled entries and 4 distinct URL sets; different slugs retain 8 distinct sets; username-only, display-only, and both-null inputs produce valid 4/4/0 matrices. #8367 deploy SHA and positive verification-run evidence are required by the environment helper, inserted into the session-local attestation, and checked again by SQL. Current OG routes plus the historical `opengraph-image-:hash` form and wildcard purge pattern are present in both external and durable inventory, with exact matrix validation. The incremental diff remains migration/test/runbook-only; no purge, schema-drop, UI, workflow, merge, or production execution was added or performed.

Checks: `git diff --check` passed; focused Vitest passed (2 files, 15 tests); targeted ESLint passed; `pnpm db:check` passed; direct null-shape contract exercise passed. `pnpm typecheck` passes on base `52d4fa3` but fails on this head with the errors called out inline.

File: apps/web/scripts/watchlist-visibility-migration.ts (hunk 4 above)
Comment (viktor-shcherb) [reply apps/web/scripts/watchlist-visibility-migration.ts]: Fixed in a48fff531. Query helpers invoked inside `sql.begin` now accept the concrete `TransactionSql` interface exported by postgres.js; only command entry points that call `.begin()` remain typed as full `Sql`. No union widening, casts, or `any` were introduced. `pnpm --filter @jobseek/web typecheck` now passes (`next typegen && tsc`), along with the focused 2-file/16-test suite, db check, targeted lint, and diff check.

File: apps/web/scripts/watchlist-visibility-contract.ts (hunk 5 above)
Comment (viktor-shcherb) [reply apps/web/scripts/watchlist-visibility-contract.ts:186]: Fixed in a48fff531. `requiredNonBlankString` now proves the environment value is a string, rejects blank input, trims it, and returns a definite `string`; the attestation interpolation therefore receives `string`, not `string | undefined`. The contract test assigns the returned value to an explicit `string`, checks trimming, and checks blank rejection. Full web typecheck passes.

**Comment by viktor-shcherb** [comment]
Rebased this draft onto current `main` after #8376 squash-merged. Because `main` now contains `0088_notification_policy_foundation`, this migration is renumbered to `0089_make_existing_watchlists_private`; the journal guard and operator confirmations were updated accordingly. The incremental diff contains no #8376 implementation files, and #8368 human approval remains a hard rollout dependency. No production migration command was run.

**Review by viktor-shcherb — COMMENTED** [review]
Technical approval for draft head `35d4ab0`; no actionable findings in the final re-review. GitHub cannot record an APPROVED review from the PR author, so this COMMENT review is the technical approval record only. It does not authorize merge, ready-for-review, deployment, or migration execution.

Verified against `main` at `bf297095`:

- The transaction-helper fix imports and uses `TransactionSql` for helpers invoked inside `sql.begin`; it does not cast transaction clients to `Sql` or otherwise bypass the type system. The prior TS2345 failures are gone.
- `WATCHLIST_ROUTE_CUTOVER_APPROVED_BY` now flows through `requiredNonBlankString`, returning a trimmed definite `string`; its compile-time assignment and blank-input rejection are tested. The prior TS2769 failure is gone.
- The merged notification migration remains `0088_notification_policy_foundation` at journal index 76. This PR appends `0089_make_existing_watchlists_private` at index 77 with a later timestamp. The runner targets 0089, requires 0088 as its prerequisite, and SQL pins the exact on-disk 0088 SHA (`c637d066...`), timestamp, and post-0088 ledger count. The notification test now locates its own journal entry instead of incorrectly requiring 0088 to remain last.
- The rebase diff is limited to the 0089 migration, targeted runner/contract, package command, migration tests, the necessary 0088 journal-test adjustment, and the runbook/docs index. No UI, purge, schema-drop, workflow, or unrelated product behavior entered the diff.
- Earlier fixes remain: labelled and distinct URL matrices cover mirrored, different, and nullable owner slugs; #8367 deploy SHA plus positive verification-run evidence are required in the runner and SQL attestation; current OG, historical `opengraph-image-:hash`, and wildcard purge forms are retained for #8370.

Checks passed: `git diff --check`; focused Vitest (3 files, 23 tests); targeted ESLint; `pnpm db:check`; and full web `pnpm typecheck` (`next typegen && tsc`). No migration or production database command was executed.

The PR must remain draft until the documented rollout dependencies are satisfied. In particular, #8368 is still open and its explicit human UI approval/deploy verification remains mandatory; this technical approval does not supersede that gate or the current repository deploy-gate status.

**Comment by viktor-shcherb** [comment]
## Watchdog rollout review — P1 evidence gap

Keep this PR Draft and dependency-blocked. A current-main synthetic merge is mechanically clean and passed the existing 23 focused tests, `pnpm db:check`, and `pnpm typecheck`, but the exact PR head has not run the normal full CI/CodeQL set and the current tests do not execute `0089_make_existing_watchlists_private` or its runner against PostgreSQL.

Before ready-for-review after #8385 is human-approved, deployed, and verified:

- rebase onto the approved current main and run full required CI;
- execute and record an ephemeral PostgreSQL inventory -> apply -> verify -> rollback round trip;
- cover zero-public rows, public rows with filters/companies/alerts, copied/source-linked watchlists, nullable owner/usernames, mirrored/equal slugs, digest stability, idempotent verification, and rollback limited to inventoried IDs;
- obtain independent database and privacy review on the exact head.

Before production execution additionally require fresh backup/restore proof, the exact immutable inventory digest, immutable deployed SHA and positive route/API probes, a named migration reviewer, a named rollback owner, and an observation-window go/no-go. No agent may merge or execute the production migration.
